testng-core/src/test/java/test/hook/HookableTest.java

```diff
@@ -1,127 +1,211 @@
-    verifyTests("Passed", new String[] {"verify"}, getPassedTests());
-    Assert.assertTrue(HookSuccess599Test.m_hook);
-    Assert.assertTrue(HookSuccess599Test.m_testWasRun);
-    Assert.assertEquals(HookSuccess599Test.m_parameter, "foo");
-  }
-
-  @Test(description = "https://github.com/cbeust/testng/pull/862")
-  public void issue862() {
-    addClass(HookSuccess862Test.class);
-    run();
-
-    verifyTests("Passed", new String[] {"verify"}, getPassedTests());
+  @DataProvider(name = "getTestClasses")
+  public Object[][] getTestClasses() {
+    return new Object[][] {
+      {HookSuccessSample.class, "Happy Flow", true},
+      {HookSuccessTimeoutSample.class, "With Timeouts (GITHUB-599)", true},
+      {HookSuccessDynamicParametersSample.class, "With Dynamic Parameters (GITHUB-862)", false}
+    };
   }
 
   @Test
   public void hookSuccessWithListener() {
-    addClass(HookSuccessWithListenerTest.class);
-    run();
-
-    verifyTests("Passed", new String[] {"verify"}, getPassedTests());
-    Assert.assertTrue(HookListener.m_hook);
+    TestNG tng = create(HookSuccessWithListenerSample.class);
+    TestResultsCollector listener = new TestResultsCollector();
+    tng.addListener(listener);
+    tng.run();
+    assertThat(listener.getPassedMethodNames()).contains("verify");
+    assertThat(listener.getPassed().get(0).getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull();
   }
 
   @Test
   public void hookFailure() {
-    addClass(HookFailureTest.class);
-    run();
-
-    // To investigate: TestNG still thinks the test passed since it can't know whether
-    // the hook ended up invoking the test or not.
-    //    verifyTests("Passed", new String[] { }, getPassedTests());
-    Assert.assertTrue(HookFailureTest.m_hook);
-    Assert.assertFalse(HookFailureTest.m_testWasRun);
+    TestNG tng = create(HookFailureSample.class);
+    TestResultsCollector listener = new TestResultsCollector();
+    tng.addListener(listener);
+    tng.run();
+    assertThat(listener.getPassedMethodNames()).isEmpty();
+    SoftAssertions assertions = new SoftAssertions();
+    listener
+        .getInvoked()
+        .forEach(
+            each -> {
+              assertions.assertThat(each.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull();
+              assertions.assertThat(each.getAttribute(HOOK_METHOD_INVOKED_ATTRIBUTE)).isNull();
+            });
+    assertions.assertAll();
   }
 
-  @Test
-  public void configurableSuccess() {
-    addClass(ConfigurableSuccessTest.class);
-    run();
-
-    Assert.assertEquals(BaseConfigurable.m_hookCount, 4);
-    Assert.assertTrue(BaseConfigurable.m_bc);
-    Assert.assertTrue(BaseConfigurable.m_bm);
-    Assert.assertTrue(BaseConfigurable.m_bs);
-    Assert.assertTrue(BaseConfigurable.m_bt);
-    Assert.assertEquals(BaseConfigurable.m_methodName, "hookWasRun");
+  @Test(dataProvider = "getConfigClasses")
+  public void configurableSuccess(Class<?> clazz, String flow) {
+    Reporter.log("Running scenario " + flow, true);
```

Comment:
Remove log

Reply:
It was kept intentionally because I squashed a bunch of independent tests into a data driven test. So that is needed.